tests/cases/compiler/contextualSignatureWithExtraParameters.ts

```diff
@@ -0,0 +1,77 @@
+// @strict: true
+// @noEmit: true
+
+// https://github.com/microsoft/TypeScript/issues/59309
+function f1(
+    cb: ((item: number) => void) | ((item: number, extra: string) => void),
+  ) {}
+  
+f1((item) => {});
+
+function f2<T>(
+    arr: T[],
+    cb: ((item: T) => void) | ((item: T, extra: unknown) => void),
+  ) {}
+  
+f2([1, 2, 3], (item) => {});
+
+export interface AsyncResultCallback<T, E = Error> {
+    (err?: E | null, result?: T): void;
+}
+
+export interface AsyncResultIterator<T, R, E = Error> {
+    (item: T, callback: AsyncResultCallback<R, E>): void;
+}
+export interface AsyncResultIteratorPromise<T, R> {
+    (item: T): Promise<R>;
+}
+
+declare function mapLimit<T, R, E = Error>(
+    arr: T[],
+    limit: number,
+    iterator:  AsyncResultIteratorPromise<T, R> | AsyncResultIterator<T, R, E>,
+): Promise<R[]>;
+
+mapLimit([1,2,3], 3, async (n) => {
+    return n ** 2;
+});
+
+function f3(
+    cb: ((a: string, b?: string) => void) | ((a: string, ...rest: string[]) => void)
+  ) {}
+  
+f3((a) => {});
+f3((a, b) => {});
+
+function f4(
+    cb: ((a: string, b: string) => void) | ((a: string, ...rest: string[]) => void)
+  ) {}
+  
+f4((a) => {});
+f4((a, b) => {});
+
+function f5(
+    cb: ((a: string, b: string) => void) | ((...rest: string[]) => void)
+  ) {}
+  
+f5((a) => {});
+f5((a, b) => {});
+
+function f6(
+  cb: ((a: string) => string) | ((a: number) => Promise<number>)
+) {}

```

Comment:
in a case like this, `cb` is not callable inside the function's body (it expects an argument of type `never`). That doesn't have to prevent contextual parameter typing since the manually-annotated parameter is still OK:
```ts
function f6(cb: ((a: string) => string) | ((a: number) => Promise<number>)) {
  cb(1)
}

f6(async (a: number) => {
  return 42;
});
```

(I'm just trying to defend (from myself) my previous suggestion to discriminate those contextual signatures based on the function's modifiers etc 😉 )

Reply:
Yeah the fact that it's not callable without further narrowing doesn't mean it's illegal to pass such a function.